### PR TITLE
Record dropped spans

### LINKED ISSUE #4156: Report dropped spans in client reports
### Description

Meta issue: https://github.com/getsentry/team-sdks/issues/88

We should also make sure to use this for when we drop spans via `beforeSendSpan`.

### PR DESCRIPTION
**Pull request by denrase**
## :bulb: Motivation and Context

Closes #4156

## :green_heart: How did you test it?

- Unit Tests

## :pencil: Checklist

You have to check all boxes before merging:

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [ ] I updated the docs if needed.
- [ ] Review from the native team if needed.
- [x] No breaking change or entry added to the changelog.
- [x] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

## :crystal_ball: Next steps

Please check if these are all instances where we need to record dropped spans, as they are a bit different from Flutter/Java and i'm not too familiar with the Cocoa SDK. 🙇

### CODE REVIEW

File: Sources/Sentry/SentryClient.m [hunk 1]
```diff
@@ -719,13 +726,35 @@ - (SentryEvent *_Nullable)prepareEvent:(SentryEvent *)event
         event.user.ipAddress = @"{{auto}}";
     }
 
+    BOOL eventIsATransaction
+        = event.type != nil && [event.type isEqualToString:SentryEnvelopeItemTypeTransaction];
+    BOOL eventIsATransactionClass
+        = eventIsATransaction && [event isKindOfClass:[SentryTransaction class]];
```
Comment (brustolin) [inline Sources/Sentry/SentryClient.m:732]: m: We already check if the event is a transaction before
```suggestion
    BOOL eventIsATransactionClass
        = !eventIsNotATransaction && [event isKindOfClass:[SentryTransaction class]];
```

File: Sources/Sentry/SentryDataCategoryMapper.m [hunk 2]
```diff
@@ -47,6 +48,9 @@
     if ([itemType isEqualToString:SentryEnvelopeItemTypeStatsd]) {
         return kSentryDataCategoryMetricBucket;
     }
+
+    // TODO: Mapping needed here?
```
Comment (brustolin) [inline Sources/Sentry/SentryDataCategoryMapper.m:52]: h: Yes

File: Sources/Sentry/include/SentryDataCategory.h [hunk 3]
```diff
@@ -17,5 +17,6 @@ typedef NS_ENUM(NSUInteger, SentryDataCategory) {
     kSentryDataCategoryMetricBucket = 8,
     kSentryDataCategoryReplay = 9,
     kSentryDataCategoryProfileChunk = 10,
-    kSentryDataCategoryUnknown = 11
+    kSentryDataCategoryUnknown = 11,
+    kSentryDataCategorySpan = 12,
```
Comment (brustolin) [inline Sources/Sentry/include/SentryDataCategory.h:21]:
```suggestion
   kSentryDataCategorySpan = 11,
   kSentryDataCategoryUnknown = 12,
```

**Review by brustolin — COMMENTED** [review]
First pass. I will wait until the PR is open to review it again.

File: Sources/Sentry/SentryClient.m (hunk 1 above)
Comment (denrase) [reply Sources/Sentry/SentryClient.m:732]: I had a crash with one of the tests. We initialize an Event, and assign it the transaction type. I was not sure if this was intetnded.

File: Sources/Sentry/include/SentryDataCategory.h (hunk 3 above)
Comment (denrase) [reply Sources/Sentry/include/SentryDataCategory.h:21]: k, so it does not matter if the raw value changes for unknown

File: Sources/Sentry/SentryDataCategoryMapper.m (hunk 2 above)
Comment (brustolin) [reply Sources/Sentry/SentryDataCategoryMapper.m]: True, span has no envelope type

**Comment by denrase** [comment]
@brustolin Do you have an idea why the benchmarking task is failing? Building with fastlane seems to be the issue?

**Comment by brustolin** [comment]
> Do you have an idea why the benchmarking task is failing?

It seems flaky.

**Comment by brustolin** [comment]
@denrase What's keeping us from merging this?

**Comment by denrase** [comment]
@brustolin Just the CI, thx for clarification 🙇

**Pull request merged**